fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/ImplementGroupByRule.java

```diff
@@ -0,0 +1,101 @@
+import com.apple.foundationdb.record.query.plan.cascades.PlanPartition;
+import com.apple.foundationdb.record.query.plan.cascades.PlannerRule;
+import com.apple.foundationdb.record.query.plan.cascades.PlannerRuleCall;
+import com.apple.foundationdb.record.query.plan.cascades.Quantifier;
+import com.apple.foundationdb.record.query.plan.cascades.RequestedOrdering;
+import com.apple.foundationdb.record.query.plan.cascades.RequestedOrderingConstraint;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.GroupByExpression;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.cascades.matching.structure.BindingMatcher;
+import com.apple.foundationdb.record.query.plan.cascades.matching.structure.ReferenceMatchers;
+import com.apple.foundationdb.record.query.plan.cascades.properties.OrderingProperty;
+import com.apple.foundationdb.record.query.plan.cascades.values.AggregateValue;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryStreamingAggregationPlan;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import javax.annotation.Nonnull;
+import java.util.Set;
+
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.AnyMatcher.any;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.QuantifierMatchers.forEachQuantifierOverRef;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.groupByExpression;
+
+/**
+ * Rule for implementing logical {@code GROUP BY} into a physical streaming aggregate operator {@link RecordQueryStreamingAggregationPlan}.
+ */
+public class ImplementGroupByRule extends PlannerRule<GroupByExpression> {
+
+    @Nonnull
+    private static final BindingMatcher<ExpressionRef<? extends RelationalExpression>> lowerRefMatcher = ReferenceMatchers.anyRef();
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> innerQuantifierMatcher = forEachQuantifierOverRef(lowerRefMatcher);
+    @Nonnull
+    private static final BindingMatcher<GroupByExpression> root =
+            groupByExpression(any(innerQuantifierMatcher));
+
+    public ImplementGroupByRule() {
+        super(root, ImmutableSet.of(RequestedOrderingConstraint.REQUESTED_ORDERING));
+    }
+
+    @Override
+    public void onMatch(@Nonnull final PlannerRuleCall call) {
+        final var bindings = call.getBindings();
+
+        final var groupByExpression = bindings.get(root);
+        final var requestedOrderings = Set.of(groupByExpression.getOrderingRequirement());
+        final var innerQuantifier = Iterables.getOnlyElement(groupByExpression.getQuantifiers());
+        final var innerReference = innerQuantifier.getRangesOver();
+        final var planPartitions = PlanPartition.rollUpTo(innerReference.getPlanPartitions(), OrderingProperty.ORDERING);
```

Comment:
You can put this into the pre of the rule so you don't have to iterate over it. We sometimes do it for other reasons but I think in your case it's perfectly fine to use an approach as e.g. in `ImplementIntersectionRule`.

Reply:
I did some refactoring to handle the case where a `GroupByExpression`'s `groupingValue` is `null` causing it (now `Optional) `getOrderingRequirement` to become `empty`.

In this case, I made it such that we `yield` for each `planPartition` since we do not really care about ordering anymore as we want to put all tuples in a single large group.

In light of that, I kept the logic as-is, could you please check it again? I am interested whether we can still pull this part and put it in the pre-part of the rule.